keps/sig-node/1287-in-place-update-pod-resources/README.md

```diff
@@ -1179,3 +1415,61 @@ information to express the idea and why it was not acceptable.
+
+### Allocated Resources
+
+If we need allocated resources & limits in the pod status API, the following options have been
+considered:
+
+**Option 1: New field "AcceptedResources"**
+
+We can't change the type of the existing field, so instead we
+introduce a new field `ContainerStatus.AcceptedResources` of type `ResourceRequirements`, to track both
+allocated requests & limits, and remove the old `AllocatedResources` field. For consistency, we also
+add `AcceptedResourcesStatus` and remove `AllocatedResourcesStatus`.
+
+Pros:
+- Consistent type across PodSpec.Container.Resources (desired), ContainerStatus.AcceptedResources (allocated), and ContainerStatus.Resources (actual)
+- If/when we implement in-place resize for DRA resources, Claims are already included in the API.
+- No need for local checkpointing, if the Kubelet can read back from the status API.
+
+Cons:
+- No path to beta without waiting a release (new fields need to start in alpha)
+- Extra code churn to migrate to the new fields
+- Inconsistent with PVC API (which has AllocatedResources), and the Node Allocatable resources.
+- The Claims field is currently unnecessary, and needs its behavior defined.
+
+Variations:
+- Use an alternative type that is a subset of the ResourceRequirements type without Claims, adding back Claims only when needed.
+- Field name ContainerStatus.Allocated, as a struct holding both the allocated resources and and the allocated resource status
+
+**Option 2: New field "AllocatedResourceLimits"**
+
+Rather than changing the type with a new field, we could use a flattened API structure and just add
+`ContainerStatus.AllocatedResourceLimits` alongside `AllocatedResources` (requests).
+
+Pros:
+- Preserves the "Allocated" name
+- Less churn to implement
+- Does not prematurely import Claims into the problem space
+
+Cons:
+- Uglier API: unnested fields adds noise to the documentation and makes it harder for humans to read the status.
+- Inconsistent types between Allocated* and Resources
+- We will want to mirror the same structure in the PodStatus for pod-level resources, and may eventually want to add AllocatedResourceClaims for DRA resource resize
+
+**Option 3: Pod-level "AllocatedResources", drop container-level API**
+
+If we assume that outside the node, controllers and people only care about pod-level allocated
+resources, then we could drop the container-level allocated resources, and just add a
+`PodStatus.AllocatedResources` field of type `ResourceRequirements`. The Kubelet still needs to track
+container-level allocation, and would use a checkpoint to do so.
+
+Pros:
+- Minimalist API, without unnecessary or redundant information
+- Preserve the "Allocated" name while still getting the advantages of type consistency
+- Similar path to beta as Option 2
+
+Cons:
+- Requires long-term checkpointing to track container allocation
+- Extra risk in assuming nothing outside the node ever needs to know container-level allocated resources.
```

Comment:
I am super +1 on approach 3. But on thinking more deeply about this, it occurred to me one thing we will lose by removing container-level AllocatedResources is the ability for o11y tools to report on actual per-container resource allocations. This could be a deal breaker.

Reply:
Container-level AllocatedResources is important also to have InPlacePodVerticalScaling alongside CPU Static Policy. We might even need to add one more container-level information like AllocatedCPUs to ensure that the CPUs critical for application performance are retained during scale down ( for example DPDK pinned ). This could be a deal breaker here to.

Reply:
> one thing we will lose by removing container-level AllocatedResources is the ability for o11y tools to report on actual per-container resource allocations. This could be a deal breaker.

What is the use case for this? It seems like desired resources & actual resources are far more useful.

> Container-level AllocatedResources is important also to have InPlacePodVerticalScaling alongside CPU Static Policy. This could be a deal breaker here to.

We will still have container level allocated resources, but it will be internal to the Kubelet. Does CPU static policy need access to allocated resources in the pod API? This is part of the reason for wanting to hold back the allocated resources API in alpha, since CPU static polciy is also out of scope for beta. If it turns out they're both needed, we can add them together.

Reply:
It is probably a "me problem" but trusting values from runtime that could be inaccurate is something I'm finding hard to swallow. This value would be picked up by o11y tools and if this is not reliable or consistently available, there's no way to know how the tools will behave (or misbehave). Let me ask around and see what the o11y folks  have to say about it. 

Also, IIRC Container.Requests are used by VPA as well (with plans to resize limits as well). If they read erroneous values, it might throw it off. cc: @gjtempleton @maciekpytel @mwielgus to chime in on this from VPA perspective.

Reply:
> We will still have container level allocated resources, but it will be internal to the Kubelet. 

Sounds good, container level allocated resources internal to kubelet should be sufficient.

> Does CPU static policy need access to allocated resources in the pod API? 

Will double check and come back.

> This is part of the reason for wanting to hold back the allocated resources API in alpha, since CPU static polciy is also out of scope for beta. If it turns out they're both needed, we can add them together.

Makes sense

Reply:
@tallclair can you capture what discussed into the pros and cons? 

@vinaykul the extra risk listed here is coming from me. If we really move forward with the task / container hierarchical scheduling, we might still need those container-level allocated resource outside the node.

Reply:
Ack. I have not had the chance to dive deeper into dynamic containers, and it may have real need. I'll try to read up more on it ... sorry my main focus has been elsewhere.

Reply:
Done.